src/fastpath/fastpath.c

```diff
@@ -15,6 +15,268 @@
 #endif
 #include <benchmark/benchmark_utilisation.h>
 
+#ifdef CONFIG_ARCH_ARM
+static inline
+#ifndef CONFIG_ARCH_ARM_V6
+FORCE_INLINE
+#endif
+#endif
+void NORETURN fastpath_signal(word_t cptr, word_t msgInfo)
+{
+    cap_t newVTable;
+    vspace_root_t *cap_pd;
+    pde_t stored_hw_asid;
+    word_t fault_type;
+    dom_t dom;
+
+    /* Get fault type. */
+    fault_type = seL4_Fault_get_seL4_FaultType(NODE_STATE(ksCurThread)->tcbFault);
+
+    /* Check there's no saved fault */
+    if (unlikely(fault_type != seL4_Fault_NullFault)) {
+        slowpath(SysSend);
+    }
+
+    /* We land up in here on every SYSCALL_SEND, so we need to check that we are
+     * actually signalling a notification */
+
+    /* Lookup the cap */
+    cap_t cap = lookup_fp(TCB_PTR_CTE_PTR(NODE_STATE(ksCurThread), tcbCTable)->cap, cptr);
+
+    /* Check it's a notification */
+    if (!cap_capType_equals(cap, cap_notification_cap)) {
+        slowpath(SysSend);
+    }
+
+    /* Check that we are allowed to signal this notification */
+    if (!unlikely(cap_notification_cap_get_capNtfnCanSend(cap))) {
+        slowpath(SysSend);
+    }
+
+    /* Get the notification address */
+    notification_t *ntfnPtr = NTFN_PTR(cap_notification_cap_get_capNtfnPtr(cap));
+
+    /* Get the notification state */
+    uint32_t ntfnState = notification_ptr_get_state(ntfnPtr);
+
+    /* Get the notification badge */
+    word_t badge = cap_notification_cap_get_capNtfnBadge(cap);
+
+    switch (ntfnState) {
+        case NtfnState_Active: {
+            word_t badge2 = notification_ptr_get_ntfnMsgIdentifier(ntfnPtr);
+            badge2 |= badge;
+            notification_ptr_set_ntfnMsgIdentifier(ntfnPtr, badge2);
+            restore_user_context();
+        }
+        case NtfnState_Idle: {
+            /* TODO: If bound thread is higher prio, we should switch to it
+             * directly via switchToThread_fp; fastpath_restore; We shouldn't
+             * need to invoke the scheduler. */
+            tcb_t *tcb = (tcb_t *)notification_ptr_get_ntfnBoundTCB(ntfnPtr);
+            /* Check if we are bound and that thread is waiting for a message */
+            if (tcb) {
```

Comment:
What values can this take?

Reply:
As in what values can `notification_ptr_get_ntfnBoundTCB` return? I presume either `0` or a valid `tcb_t` address. Some other usages of `notification_ptr_get_ntfnBoundTCB` seem to use it the same way e.g. [here](https://github.com/seL4/seL4/blob/master/src/object/notification.c#L80-L83) and [here](https://github.com/seL4/seL4/blob/master/src/object/notification.c#L277-L281).